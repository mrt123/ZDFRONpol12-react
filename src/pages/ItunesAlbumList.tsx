import { useEffect, useState } from "react";
import ItunesAlbum from "../components/ItunesAlbum";

interface AlbumDataEntry {
  ["im:name"]: {
    label: string;
  };
}

interface ItunesTopAlbumsResponseData {
  feed: {
    entry: AlbumDataEntry[];
  };
}

const ItunesAlbumList = () => {
  const [albumDataEntries, setAlbumDataEntries] = useState<
    AlbumDataEntry[] | null
  >(null);

  useEffect(() => {
    const fetchAndSetAlbums = async () => {
      const fetchResult = await fetch(
        "https://itunes.apple.com/us/rss/topalbums/limit=100/json"
      );

      const fetchedData =
        (await fetchResult.json()) as ItunesTopAlbumsResponseData;

      console.log({ fetchedData });
      setAlbumDataEntries(fetchedData.feed.entry);
    };

    fetchAndSetAlbums();
  }, []);

  const albumComponents = albumDataEntries?.map((albumEntry, index) => {
    return <ItunesAlbum key={index} albumDataEntry={albumEntry} />;
  });

  return (
    <>
      <div title="Hey I'm a div">Top Albums</div>
      {albumComponents}
    </>
  );
};

export default ItunesAlbumList;
