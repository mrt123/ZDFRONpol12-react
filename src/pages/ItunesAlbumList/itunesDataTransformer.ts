import { AlbumProps } from "../../common/musicAlbum/Album";

interface ObjectWithlabel {
  label: string;
}

interface PriceObject {
  label: string;
  attributes: {
    amount: string;
  };
}

export interface ItunesAlbumDataEntry {
  "im:name": ObjectWithlabel;
  "im:artist": ObjectWithlabel;
  "im:price"?: PriceObject;
  "im:image"?: ObjectWithlabel[];
}

export const getAlbumsFromItunesAlbumData = (
  itunesAlbumEntries: ItunesAlbumDataEntry[]
): AlbumProps[] => {
  const transformedData = itunesAlbumEntries.map((itunesAlbum, index) => {
    const priceValueFromItunes = itunesAlbum?.["im:price"]?.attributes.amount;

    return {
      number: index + 1,
      coverImageUrl: itunesAlbum?.["im:image"]?.[1]?.label || "",
      description: {
        title: itunesAlbum["im:name"].label,
        artist: itunesAlbum["im:artist"].label,
        price: Number(priceValueFromItunes),
      },
    };
  });

  return transformedData;
};
