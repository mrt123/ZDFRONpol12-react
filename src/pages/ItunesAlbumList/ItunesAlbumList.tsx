import { useEffect, useState } from "react";
import Album from "../../common/album/Album";
import CircleLoader from "../../common/CircleLoader";
import { ItunesAlbumDataEntry } from "../../common/ItunesAlbum";
import PageTitle from "../../common/PageTitle";
import { getAlbumsFromItunesAlbumData } from "./itunesDataTransformer";

const addDelay = () => {
  return new Promise((resolve) =>
    setTimeout(() => {
      resolve(null);
    }, 1000)
  );
};

interface ItunesTopAlbumsResponseData {
  feed: {
    entry: ItunesAlbumDataEntry[];
  };
}

const ItunesAlbumList = () => {
  const [albumDataEntries, setAlbumDataEntries] = useState<
    ItunesAlbumDataEntry[] | null
  >(null);

  useEffect(() => {
    const fetchAndSetAlbums = async () => {
      const fetchResponse = await fetch(
        "https://itunes.apple.com/us/rss/topalbums/limit=100/json"
      );

      await addDelay();

      const fetchedData =
        (await fetchResponse.json()) as ItunesTopAlbumsResponseData;

      console.log({ fetchedData });
      setAlbumDataEntries(fetchedData.feed.entry);
    };

    fetchAndSetAlbums();
  }, []);

  const albumData = albumDataEntries
    ? getAlbumsFromItunesAlbumData(albumDataEntries)
    : [];

  const albumComponents = albumData?.map((album) => {
    return (
      <Album
        number={album.number}
        coverImageUrl={album.coverImageUrl}
        description={album.description}
      />
    );
  });

  return (
    <>
      <PageTitle title="Top Albums" />
      <CircleLoader show={albumDataEntries === null} />
      {albumComponents}
    </>
  );
};

export default ItunesAlbumList;
