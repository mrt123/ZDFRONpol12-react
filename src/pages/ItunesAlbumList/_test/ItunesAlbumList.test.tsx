import { render, waitFor } from "@testing-library/react";
import ItunesAlbumList from "../ItunesAlbumList";
import { ItunesAlbumDataEntry } from "../itunesDataTransformer";

describe("ItunesAlbumList", () => {
  it("matches snapshot when data is loading", () => {
    const { container } = render(<ItunesAlbumList />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot after data is loaded", async () => {
    // ARRANGE
    const rawAlbumEntires: ItunesAlbumDataEntry[] = [
      {
        "im:name": {
          label: "Album1",
        },
        "im:artist": {
          label: "Michael Jackson",
        },
        "im:releaseDate": {
          label: "1999/05/01",
        },
        "im:price": {
          attributes: {
            amount: "99000000",
          },
          label: "99000000",
        },
      },
      {
        "im:name": {
          label: "Killing me softly",
        },
        "im:artist": {
          label: "Fugees",
        },
        "im:releaseDate": {
          label: "1996/05/27",
        },
      },
    ];

    const rawAlbumTestData = {
      feed: {
        entry: rawAlbumEntires,
      },
    };

    window.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve(rawAlbumTestData),
      })
    ) as jest.Mock;

    // ACT
    const { container } = render(<ItunesAlbumList />);

    const loaderElements = container.getElementsByClassName("loader");

    await waitFor(
      () => {
        return expect(loaderElements.length === 0).toEqual(true);
      },
      {
        timeout: 13000,
      }
    );

    // ASSERT
    // We do not test snapshot here, we avoid complex snapshot comparison. Imagine this is the most complex component in a large application.
    expect(container).toMatchSnapshot();
  });
});
