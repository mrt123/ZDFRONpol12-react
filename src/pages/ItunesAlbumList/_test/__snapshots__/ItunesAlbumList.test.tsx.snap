// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItunesAlbumList matches snapshot after data is loaded 1`] = `
<div>
  <div
    style="margin-bottom: 20px; text-align: center;"
  >
    Top Albums
  </div>
  <div
    style="display: flex; margin: 20px; font-size: 12px; background: rgb(53, 58, 69);"
  >
    <div
      style="padding: 5px; background: black;"
    >
      1
    </div>
    <div
      style="width: 100px; height: 100px; background: blue; background-image: url(); background-size: contain;"
    >
       
    </div>
    <div
      style="margin: 8px;"
    >
      <div
        style="margin: 5px; font-size: 11px;"
      >
        Album1
      </div>
      <div
        style="margin: 5px; font-size: 11px;"
      >
        Michael Jackson
      </div>
      <div
        style="margin: 5px; font-size: 11px;"
      >
        99000000
      </div>
      <div
        style="margin: 5px; font-size: 11px;"
      >
        released: 8624 days ago
      </div>
    </div>
  </div>
  <div
    style="display: flex; margin: 20px; font-size: 12px; background: rgb(53, 58, 69);"
  >
    <div
      style="padding: 5px; background: black;"
    >
      2
    </div>
    <div
      style="width: 100px; height: 100px; background: blue; background-image: url(); background-size: contain;"
    >
       
    </div>
    <div
      style="margin: 8px;"
    >
      <div
        style="margin: 5px; font-size: 11px;"
      >
        Killing me softly
      </div>
      <div
        style="margin: 5px; font-size: 11px;"
      >
        Fugees
      </div>
      <div
        style="margin: 5px; font-size: 11px;"
      >
        N/A
      </div>
      <div
        style="margin: 5px; font-size: 11px;"
      >
        released: 9693 days ago
      </div>
    </div>
  </div>
</div>
`;

exports[`ItunesAlbumList matches snapshot when data is loading 1`] = `
<div>
  <div
    style="margin-bottom: 20px; text-align: center;"
  >
    Top Albums
  </div>
  <div
    style="text-align: center; margin-top: 100px;"
  >
    <div
      class="loader"
    />
  </div>
</div>
`;
